File: gateway/go/internal/risk/budget.go
// SPDX-License-Identifier: AGPL-3.0-or-later
// Commercial licence available
// © Concepts 1996–2026 Miroslav Šotek. All rights reserved.
// © Code 2020–2026 Miroslav Šotek. All rights reserved.
// ORCID: 0009-0009-3560-0851
// Contact: www.anulum.li | protoscience@anulum.li
// Director-Class AI — per-tenant sliding-window risk budget

package risk

import (
	"container/list"
	"sync"
	"time"
)

// BudgetEntry mirrors the Python ``BudgetEntry`` dataclass.
// ``Accepted`` tells the caller whether the last reservation
// succeeded — without it, a non-zero ``Remaining`` could reflect
// either "reservation applied" or "reservation refused, ledger
// untouched".
type BudgetEntry struct {
	TenantID       string
	WindowSeconds  float64
	Allowance      float64
	Consumed       float64
	Remaining      float64
	Events         int
	Accepted       bool
}

// Exhausted reports whether the caller should reject the request.
func (e BudgetEntry) Exhausted() bool {
	return !e.Accepted || e.Remaining <= 0.0
}

// Budget is a sliding-window risk budget. Safe for concurrent use.
type Budget struct {
	allowance     float64
	window        time.Duration
	overrides     map[string]float64
	ledgers       map[string]*list.List
	totals        map[string]float64
	clock         func() time.Time
	mu            sync.Mutex
}

// NewBudget builds a budget with ``allowance`` per ``window``.
// Callers pass ``nil`` for ``clock`` to use ``time.Now``.
func NewBudget(allowance, windowSeconds float64, clock func() time.Time) (*Budget, error) {
	if allowance <= 0 {
		return nil, &invalidParamError{field: "allowance", value: allowance}
	}
	if windowSeconds <= 0 {
		return nil, &invalidParamError{field: "windowSeconds", value: windowSeconds}
	}
	if clock == nil {
		clock = time.Now
	}
	return &Budget{
		allowance: allowance,
		window:    time.Duration(windowSeconds * float64(time.Second)),
		overrides: make(map[string]float64),
		ledgers:   make(map[string]*list.List),
		totals:    make(map[string]float64),
		clock:     clock,
	}, nil
}

// SetAllowance overrides the allowance for a single tenant.
func (b *Budget) SetAllowance(tenantID string, allowance float64) error {
	if allowance <= 0 {
		return &invalidParamError{field: "allowance", value: allowance}
	}
	b.mu.Lock()
	defer b.mu.Unlock()
	b.overrides[tenantID] = allowance
	return nil
}

// AllowanceFor returns the effective allowance for ``tenantID``.
func (b *Budget) AllowanceFor(tenantID string) float64 {
	b.mu.Lock()
	defer b.mu.Unlock()
	if a, ok := b.overrides[tenantID]; ok {
		return a
	}
	return b.allowance
}

// Reserve attempts to charge ``risk`` (clamped to [0, 1]) against
// ``tenantID``'s ledger. Pass zero to read the state without
// charging.
func (b *Budget) Reserve(tenantID string, risk float64) BudgetEntry {
	if risk < 0 {
		risk = 0
	} else if risk > 1 {
		risk = 1
	}
	now := b.clock()
	cutoff := now.Add(-b.window)
	b.mu.Lock()
	defer b.mu.Unlock()
	ledger, ok := b.ledgers[tenantID]
	if !ok {
		ledger = list.New()
		b.ledgers[tenantID] = ledger
	}
	// Prune expired entries.
	for e := ledger.Front(); e != nil; {
		entry := e.Value.(*ledgerEntry)
		if entry.ts.Before(cutoff) {
			next := e.Next()
			ledger.Remove(e)
			b.totals[tenantID] -= entry.risk
			e = next
			continue
		}
		break
	}
	allowance := b.allowance
	if a, ok := b.overrides[tenantID]; ok {
		allowance = a
	}
	tentative := b.totals[tenantID] + risk
	if tentative <= allowance+1e-9 {
		if risk > 0 {
			ledger.PushBack(&ledgerEntry{ts: now, risk: risk})
			b.totals[tenantID] += risk
		}
		return BudgetEntry{
			TenantID:      tenantID,
			WindowSeconds: b.window.Seconds(),
			Allowance:     allowance,
			Consumed:      b.totals[tenantID],
			Remaining:     allowance - b.totals[tenantID],
			Events:        ledger.Len(),
			Accepted:      true,
		}
	}
	return BudgetEntry{
		TenantID:      tenantID,
		WindowSeconds: b.window.Seconds(),
		Allowance:     allowance,
		Consumed:      b.totals[tenantID],
		Remaining:     allowance - b.totals[tenantID],
		Events:        ledger.Len(),
		Accepted:      false,
	}
}

// Snapshot reads the current ledger without charging.
func (b *Budget) Snapshot(tenantID string) BudgetEntry {
	return b.Reserve(tenantID, 0)
}

// Reset clears one tenant's ledger. Passing the empty string clears
// every tenant.
func (b *Budget) Reset(tenantID string) {
	b.mu.Lock()
	defer b.mu.Unlock()
	if tenantID == "" {
		b.ledgers = make(map[string]*list.List)
		b.totals = make(map[string]float64)
		return
	}
	delete(b.ledgers, tenantID)
	delete(b.totals, tenantID)
}

type ledgerEntry struct {
	ts   time.Time
	risk float64
}

File: src/director_ai/core/routing/budget.py
# SPDX-License-Identifier: AGPL-3.0-or-later
# Commercial license available
# © Concepts 1996–2026 Miroslav Šotek. All rights reserved.
# © Code 2020–2026 Miroslav Šotek. All rights reserved.
# ORCID: 0009-0009-3560-0851
# Contact: www.anulum.li | protoscience@anulum.li
# Director-Class AI — per-tenant risk budget

"""Sliding-window risk budget per tenant.

The :class:`PromptRiskScorer` assigns every prompt a ``[0, 1]``
risk score; the budget tracks a rolling sum of those scores per
tenant over a configurable window and refuses further reservations
once the window total exceeds the allowance.

The budget is intentionally simple: in-memory, single-process, no
persistence. A Redis-backed variant can plug in later by
implementing the same public contract — ``reserve``, ``snapshot``,
``reset`` — which is why :class:`RiskBudget` is kept free of any
Redis coupling today.
"""

from __future__ import annotations

import threading
import time
from collections import deque
from collections.abc import Callable
from dataclasses import dataclass, field


@dataclass(frozen=True)
class BudgetEntry:
    """Snapshot of one tenant's current budget state.

    ``accepted`` distinguishes "reservation charged" from "over
    budget — ledger untouched". Without this flag a caller could
    not tell whether ``remaining > 0`` meant "reservation applied"
    or "reservation denied and ledger unchanged".
    """

    tenant_id: str
    window_seconds: float
    allowance: float
    consumed: float
    remaining: float
    events: int
    accepted: bool = True

    @property
    def exhausted(self) -> bool:
        """True when the last reservation was refused or the budget
        is empty — the caller should reject the request."""
        return not self.accepted or self.remaining <= 0.0


@dataclass
class _TenantLedger:
    """Private per-tenant deque of ``(timestamp, risk)`` entries."""

    entries: deque[tuple[float, float]] = field(default_factory=deque)
    total: float = 0.0

    def prune(self, cutoff: float) -> None:
        while self.entries and self.entries[0][0] < cutoff:
            _, expired = self.entries.popleft()
            self.total -= expired
            if self.total < 0.0:  # pragma: no cover — float drift guard
                self.total = 0.0

    def push(self, timestamp: float, risk: float) -> None:
        self.entries.append((timestamp, risk))
        self.total += risk

    def count(self) -> int:
        return len(self.entries)


class RiskBudget:
    """Sliding-window risk budget with per-tenant ledgers.

    Parameters
    ----------
    allowance :
        Total risk the tenant may consume in one window.
    window_seconds :
        Window length. Events older than this are pruned on the
        next access.
    clock :
        Injectable ``() -> float`` — tests pass a fake clock to make
        the window deterministic. Defaults to :func:`time.monotonic`.
    default_allowances :
        Optional per-tenant overrides; missing tenants fall back to
        the global ``allowance``.
    """

    def __init__(
        self,
        *,
        allowance: float = 10.0,
        window_seconds: float = 60.0,
        clock: Callable[[], float] = time.monotonic,
        default_allowances: dict[str, float] | None = None,
    ) -> None:
        if allowance <= 0.0:
            raise ValueError(f"allowance must be positive; got {allowance}")
        if window_seconds <= 0.0:
            raise ValueError(
                f"window_seconds must be positive; got {window_seconds}"
            )
        self._allowance = float(allowance)
        self._window = float(window_seconds)
        self._clock = clock
        self._overrides: dict[str, float] = dict(default_allowances or {})
        self._ledgers: dict[str, _TenantLedger] = {}
        self._lock = threading.Lock()

    def set_allowance(self, tenant_id: str, allowance: float) -> None:
        """Override the allowance for one tenant."""
        if allowance <= 0.0:
            raise ValueError(f"allowance must be positive; got {allowance}")
        with self._lock:
            self._overrides[tenant_id] = float(allowance)

    def allowance_for(self, tenant_id: str) -> float:
        """Return the effective allowance for ``tenant_id``."""
        with self._lock:
            return self._overrides.get(tenant_id, self._allowance)

    def reserve(self, tenant_id: str, risk: float) -> BudgetEntry:
        """Try to reserve ``risk`` for ``tenant_id``.

        Returns a :class:`BudgetEntry` reflecting the post-decision
        state. The reservation succeeds when ``entry.remaining >= 0``
        after the charge; callers should check ``entry.exhausted``
        to decide whether to reject / queue the request.

        ``risk`` is clamped to ``[0, 1]``. Pass ``0`` to check the
        current state without consuming budget.
        """
        if risk < 0.0:
            risk = 0.0
        elif risk > 1.0:
            risk = 1.0
        now = self._clock()
        cutoff = now - self._window
        with self._lock:
            ledger = self._ledgers.setdefault(tenant_id, _TenantLedger())
            ledger.prune(cutoff)
            allowance = self._overrides.get(tenant_id, self._allowance)
            tentative = ledger.total + risk
            if tentative <= allowance + 1e-9:
                if risk > 0.0:
                    ledger.push(now, risk)
                consumed = ledger.total
                remaining = allowance - consumed
                return BudgetEntry(
                    tenant_id=tenant_id,
                    window_seconds=self._window,
                    allowance=allowance,
                    consumed=consumed,
                    remaining=remaining,
                    events=ledger.count(),
                    accepted=True,
                )
            # Over budget — report without charging.
            consumed = ledger.total
            return BudgetEntry(
                tenant_id=tenant_id,
                window_seconds=self._window,
                allowance=allowance,
                consumed=consumed,
                remaining=allowance - consumed,
                events=ledger.count(),
                accepted=False,
            )

    def snapshot(self, tenant_id: str) -> BudgetEntry:
        """Read current ledger state without charging."""
        return self.reserve(tenant_id, 0.0)

    def reset(self, tenant_id: str | None = None) -> None:
        """Clear one tenant's ledger (``tenant_id`` given) or every
        ledger at once (``tenant_id=None``). Useful for tests and
        for operator-triggered 'new epoch' events."""
        with self._lock:
            if tenant_id is None:
                self._ledgers.clear()
            else:
                self._ledgers.pop(tenant_id, None)
